server/src/main/java/org/elasticsearch/cluster/metadata/IndexReshardingMetadata.java

```diff
@@ -0,0 +1,214 @@
+ * We only allow at most a single resharding operation to be in flight for an index, so removing this metadata is a prerequisite
+ * to beginning another resharding operation.
+ */
+public class IndexReshardingMetadata implements ToXContentFragment, Writeable {
+    private final IndexReshardingState state;
+
+    private static final String SPLIT_FIELD_NAME = "split";
+    private static final ParseField SPLIT_FIELD = new ParseField(SPLIT_FIELD_NAME);
+    // This exists only so that tests can verify that IndexReshardingMetadata supports more than one kind of operation.
+    // It can be removed when we have defined a second real operation, such as shrink.
+    private static final String NOOP_FIELD_NAME = "noop";
+    private static final ParseField NOOP_FIELD = new ParseField(NOOP_FIELD_NAME);
+
+    private static final ConstructingObjectParser<IndexReshardingMetadata, Void> PARSER = new ConstructingObjectParser<>(
+        "index_resharding_metadata",
+        args -> {
+            // the parser ensures exactly one argument will not be null
+            if (args[0] != null) {
+                return new IndexReshardingMetadata((IndexReshardingState) args[0]);
+            } else {
+                return new IndexReshardingMetadata((IndexReshardingState) args[1]);
+            }
+        }
+    );
+
+    static {
+        PARSER.declareObjectOrNull(
+            ConstructingObjectParser.optionalConstructorArg(),
+            (parser, c) -> IndexReshardingState.Split.fromXContent(parser),
+            null,
+            SPLIT_FIELD
+        );
+        PARSER.declareObjectOrNull(
+            ConstructingObjectParser.optionalConstructorArg(),
+            (parser, c) -> IndexReshardingState.Noop.fromXContent(parser),
+            null,
+            NOOP_FIELD
+        );
+        PARSER.declareExclusiveFieldSet(SPLIT_FIELD.getPreferredName(), NOOP_FIELD.getPreferredName());
+        PARSER.declareRequiredFieldSet(SPLIT_FIELD.getPreferredName(), NOOP_FIELD.getPreferredName());
+    }
+
+    // for testing
+    IndexReshardingState getState() {
+        return state;
+    }
+
+    // visible for testing
+    IndexReshardingMetadata(IndexReshardingState state) {
+        this.state = state;
+    }
+
+    static IndexReshardingMetadata fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        String name = switch (state) {
+            case IndexReshardingState.Noop ignored -> NOOP_FIELD.getPreferredName();
+            case IndexReshardingState.Split ignored -> SPLIT_FIELD.getPreferredName();
+        };
+        builder.startObject(name);
+        state.toXContent(builder, params);
+        builder.endObject();
+
+        return builder;
+    }
+
+    public IndexReshardingMetadata(StreamInput in) throws IOException {
```

Comment:
We mostly put ctors at the tope of classes, below fields. So I feel like this would best go above `getState`.

Reply:
Done in 20264b8